numpy/lib/function_base.py

```diff
@@ -3792,16 +4029,206 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
+    def _normalize_and_set_boundary(ws):  # ws is a 1-D vector
+        w_vec = ws.copy()
+        inds = w_vec > 0
+        if (w_vec[inds] < 1).any():
+            w_vec[inds] = w_vec[inds] / w_vec[inds].min()
+        if len(w_vec) > 1:
+            w_vec[0] -= 1
+        return w_vec
+
+    normalize_and_bound = np.vectorize(_normalize_and_set_boundary,
+                                       signature='(i)->(i)')
+    ws_sorted = normalize_and_bound(ws_sorted)
+
+    nonzero_w_inds = ws_sorted > 0
+
+    # compute the cumulative weights
+    cum_w = ws_sorted.cumsum(axis=-1)
+    cum_w_max = cum_w.max(axis=-1, keepdims=True)
+
+    # (5) compute the upper bounds of probability bands
+    # upper bound is just the normalized cumulative weight
+    prob_band_upper_bounds = cum_w / cum_w_max
+
+    # (6) finding lower bound requires computing the transition band width,
+    # from the prior band's right edge to the current band's left edge,
+    # which is (current band upper bound - prior band upper bound) / weight
+    # first need the prior band upper bound
+    prior_band_upper_bounds = np.roll(prob_band_upper_bounds, 1, axis=-1)
+    prior_band_upper_bounds[..., 0] = 0  # nothing prior to left-most band
+
+    # the weight denominator could contain 0's (due to input or nans)
+    # so the division is only performed where weight > 0 (nonzero_w_inds)
+    transition_band_width = ws_sorted  # to initialize
+    # transition_band_width inherits the fact that
+    # band width == 0 where weight == 0
+    transition_band_width[nonzero_w_inds] =\
+        ((prob_band_upper_bounds[nonzero_w_inds] -
+          prior_band_upper_bounds[nonzero_w_inds]) /
+         ws_sorted[nonzero_w_inds])  # ok to overwrite ws_sorted
+
+    # the transition_band_width computed should align with the left band
+    # it bounds to, so it's rolled to the left by 1 position
+    transition_band_width = np.roll(transition_band_width, -1, axis=-1)
+
+    # (7) the lower bound to the current probability band is equal to
+    # prior band's upper bound + prior band's transition band width,
+    # so we sum and then roll to the right by 1 position
+    prob_band_lower_bounds =\
+        np.roll(prob_band_upper_bounds + transition_band_width, 1, axis=-1)
+    prob_band_lower_bounds[..., 0] = 0  # boundary condition on the left
+
+    # combine and resort
+    quantile_bounds = np.concatenate([prob_band_upper_bounds,
+                                      prob_band_lower_bounds],
+                                     axis=-1)
+    inds_resort = np.argsort(quantile_bounds, axis=-1)
+    quantile_bounds = arraysort(quantile_bounds, inds_resort)
+
+    ap = np.concatenate([ap_sorted, ap_sorted], axis=-1)
+    ap = arraysort(ap, inds_resort)
+
+    # (8) interpolate for indices where q sits along axis of ap
+    Nx = ap.shape[-1]
+    indices_hard = np.arange(Nx)
+    vec_interp_func = np.vectorize(np.interp, signature='(n),(m),(m)->(n)')
```

Comment:
Say I'm evaluating `np.percentile(a, q=50, axis=0, weights=[1, 1])`.  When the code path gets to the above snippet, we'd see
```
>>> indices_hard
array([0, 1, 2, 3])
>>> quantile_bounds
array([[0., 0., 1., 1.],
       [0., 0., 1., 1.],
       [0., 0., 1., 1.]])
>>> q
array([0.5])
```
This is what I currently have:
```
>>> vec_interp_func = np.vectorize(np.interp, signature='(n),(m),(m)->(n)')
>>> indices = vec_interp_func(q, quantile_bounds, indices_hard)
>>> indices
array([[1.5],
       [1.5],
       [1.5]])
>>> indices.shape
(3, 1)
>>> quantile_bounds.shape
(3, 4)
```
which would give me what I want for `x1` and `x2` when I later take the indices along axis -1 using
```
vec_take = np.vectorize(_take1d, signature='(n),(m),(m)->(n)')
```

On the other hand, changing the signature from `(n)` to `()`:
```
>>> vec_interp_func = np.vectorize(np.interp, signature='(),(m),(m)->()')
>>> indices = vec_interp_func(q, quantile_bounds, indices_hard)
>>> indices
array([1.5, 1.5, 1.5])
>>> indices.shape
(3,)
```
would lead to the wrong result.  More like triple the result.